### ISSUE TITLE
[FEATURE REQUEST] Recursively populate

### ISSUE THREAD

**Issue by pixelfreak** [balderdashy/waterline#308]
I'm on Sails 0.10. I have a Model which has a collection of Model which in turn has more Model. I know how to populate the first child but how do I recursively populate the rest of the children?

**Comment by urielaero** [comment]
+1

**Comment by luolongzhi** [comment]
+1

**Comment by mydearxym** [comment]
+1

**Comment by jaumard** [comment]
+1

**Comment by rakeshmenon** [comment]
+1

**Comment by Thelmos** [comment]
Definitely a must.

**Comment by Batornator** [comment]
+1

Any news on this feature? Would be ideal for a current project. Are there plans to switch to knex or will there be native support for a recursive populate?

**Comment by mattblang** [comment]
+1

**Comment by mattblang** [comment]
+1
Any news about this?

**Comment by liratanak** [comment]
+1

**Comment by jeromewir** [comment]
+1

Definitely need this feature !

**Comment by lghamie** [comment]
+1 Totally need this.

**Comment by mkalish** [comment]
+1

**Comment by felipecm** [comment]
+1

**Comment by besarthoxhaj** [comment]
@pixelfreak I got the same issue `res.json(theModel)` and `JSON.stringify(theModel)` won't show up the properties populated.

How did you fixed it?

**Comment by besarthoxhaj** [comment]
Ok this is how I did it:

```js
Posts
.findOne(2)
.populate('comments')
.exec(function (err, itemPost){
	// some error handling
	async.map(itemPost.comments,
		function (comment, callback){
			Comments
			.findOne(comment.id)
			.populateAll()
			.exec(function (errComments, commentItem){
				if(errComments){return callback(errComments)};
				callback(null, commentItem);
			});
		},
		function (errFromIterator, results){
			if(errFromIterator){res.serverError()};

			var itemPostToJSON = itemPost.toJSON();

			itemPostToJSON.comments = results;

			var clone = _.clone(itemPostToJSON);
			
			res.send(clone);
		}
	)
});
```

I'm sure there are better ways, so don't be shy and post them:)

**Comment by felipecm** [comment]
It's ugly.. but works with two levels and only two database calls

```javascript
    async.waterfall([
      function(callback) {
        Groups
          .find({parent:parent})
          .populate('nodes')
          .exec(function(err,data){
            callback(err, data);
          });
      },
      function(item, callback) {

        var data = _.flattenDeep(_.map(item,function(obj){
          return _.pluck(obj.nodes,'id');
        }));

        Groups.find({id: data}).exec(function(err,nodes){
          callback(err, item,nodes);
        });
      },
      function(items,nodes, callback) {

        var deepNodes = _.indexBy(nodes, 'id');

        callback(null,items.map(function(o){

          var tree = o.toObject();

          tree.nodes.map(function(node) {
            node = (deepNodes[node.id]);
            return node;
          });
          
          return _.clone(tree);
        }));
      }
    ], function (err, results) {
      
      if (err) {return res.serverError(err);}
      
      return res.json(results);
    });
```

**Comment by besarthoxhaj** [comment]
@felipecm thanks for sharing:)

Would be great to see how other people did it or how @particlebanana solved it.

**Comment by craigvl** [comment]
+1

**Comment by jakubstrama** [comment]
+1

**Comment by terminox** [comment]
+1

it's very useful

**Comment by Nicolab** [comment]
+1

**Comment by webzepter** [comment]
+1

**Comment by dmarcelino** [comment]
For those following this there is also an interesting discussion about this topic on #256 (now closed as a duplicate).

**Comment by dmarcelino** [comment]
Adding what @particlebanana said in [sails-mongo#108](https://github.com/balderdashy/sails-mongo/issues/108#issuecomment-41731379) so we can close duplicates and move conversation about this topic to this thread:

> We are working on a populate command that takes the dot notation:
> 
> ```javascript
> .populate('pet')
> .populate('pet.owner')
> ```
> 
> It's a little more complicated for sequel databases but in-memory operations for databases that don't support joins (like mongo) should be a little easier.

**Comment by pavel-anisimov** [comment]
+1. Something like that would be just perfect. 
```
User.findOne().populate('assignment').populate('assignment.location').exec();
```

**Comment by svenvarkel** [comment]
Just a thought that I had. It came 'cause I'm working with SQL and noSQL solutions in parallel. 
In context of SQL (MySQL) populate() would just mean to INNER JOIN another table by foreign keys. In context of noSQL I'm not sure what it should be exactly.

**Comment by rathahin** [comment]
+1  Model.findOne().populate('children.children.children').exec();

**Comment by stenio123** [comment]
+1

**Comment by cadrogui** [comment]
+1

**Comment by HerrZatacke** [comment]
+1

**Comment by HerrZatacke** [comment]
+1

**Comment by athk** [comment]
+1

**Comment by devinivy** [comment]
If you're interested in the future of this feature, please subscribe above– works even better than a +1!

**Comment by myusuf91** [comment]
+1

**Comment by nachiket-p** [comment]
+1

**Comment by brandonsimpson** [comment]
@dmarcelino +1 for populating deep associations

```
.populate('pet')
.populate('pet.owner')
```

I am currently forced to use `promise.props` to pull all of the various data sets and map them together after all of the results are returned. It's very clunky compared to just building a sql join statement and getting all results from the database.

**Comment by dinana** [comment]
I developed this function to handle deep population, feel free to use it:
https://gist.github.com/dinana/52453ecb00d469bb7f12

**Comment by tjwebb** [comment]
@dinana would you mind submitting a PR?

**Comment by dinana** [comment]
@tjwebb As an add-on to query/populate or as a service still?

**Comment by tjwebb** [comment]
I think it'd be great to have as a feature of waterline

**Comment by dinana** [comment]
I'll need some pointers to properly integrate this into waterline code, can you help? Also, do you use Gitter/IRC/else?

**Comment by Tenkir** [comment]
The only issue I've found with this is it does not work with more than one model. So if you want to find multiple instances you have to hack around.

**Comment by aato** [comment]
+1

**Comment by albertkim** [comment]
Indeed, I'm having to use async to map the fields I want together, but it is incredibly inelegant. +1 from me.

**Comment by Tenkir** [comment]
Just use dinana's solution. It works perfectly

**Comment by zorosun** [comment]
+1

**Comment by jaxoncreed** [comment]
+1

**Comment by atiertant** [comment]
just PR https://github.com/balderdashy/waterline/pull/1052
@dinana thanks for original idea
@Tenkir and others could you test this code please ?

**Comment by RXminuS** [comment]
Nice @atiertant!

**Comment by zorosun** [comment]
@atiertant It works for me too, dinana's solution

**Comment by Tenkir** [comment]
@atiertant I will check it out when I get out of work around 6, but it looks to be all good.

**Comment by stuk88** [comment]
+1

**Comment by baltag** [comment]
+1

**Comment by JonGarbayo** [comment]
+1

**Comment by particlebanana** [comment]
Alright guys, we know about this. If you want to stay up to date on this subscribe to the issue above.

**Comment by sailsbot** [comment]
Thanks for posting, @pixelfreak.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by dmarcelino** [comment]
Reopening as this one has PR #1052 associated with it.

**Comment by sailsbot** [comment]
Thanks for posting, @pixelfreak.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @pixelfreak.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
